src/dune/expander.ml

```diff
@@ -324,8 +351,9 @@ let expand_and_record acc ~map_exe ~dep_kind ~expansion_kind
   in
   let add_fail =
     match expansion_kind with
-    | Static -> fun _ (f : fail) -> f.fail ()
-    | Dynamic -> Resolved_forms.add_fail
+    | Static -> Or_exn.raise
```

Comment:
This wrapping is necessary so that the exception is caught by `Expander.Or_error`. We want to make sure that we don't accidentally delay exceptions that weren't raised by `add_fail`.

Reply:
I'm wondering if that's really necessary. I tried removing the wrapping and nothing broke in the testsuite. I'm also thinking that we could remove `add_fail` entirely and just add a catch all around `expand_and_record`. That would make the code simpler and easier to reason about. WDYT?

Reply:
> I'm wondering if that's really necessary. I tried removing the wrapping and nothing broke in the testsuite.

Indeed. Another option is to just catch & delay `User_error.E`. I think we can reasonably assume that all errors can be delayed.

>  I'm also thinking that we could remove add_fail entirely and just add a catch all around expand_and_record. That would make the code simpler and easier to reason about. WDYT?

Seems fine to me. Just to confirm, you would then have to handle this expansion based on the `expansion_kind`. I.e.

```ocaml
...
with User_error.E _ as exn ->
  match exception_kind with
  | Dynamic -> Resolved_forms.add acc exn
  | Static -> reraise exn
```

Reply:
Yep. But I don't even think we need to match on `User_error.E` here. Catching all exceptions seems fine. We are just making the behaviour lazy, that's all.

Reply:
I think it would be quite bad to swallow `Code_error` here. See the last commit, I've simplified the code further but I'm still not catching only User_errors. I think that's a safe compromise.

Reply:
Seems reasonable